src/core/lib/iomgr/timer.cc

```diff
@@ -29,6 +30,7 @@ void grpc_set_timer_impl(grpc_timer_vtable* vtable) {
 
 void grpc_timer_init(grpc_timer* timer, grpc_millis deadline,
                      grpc_closure* closure) {
+  grpc_core::ExecCtx::Get()->InvalidateNow();
```

Comment:
This doesn't seem like it will really help, because by the time we get here, the calling code has already used `ExecCtx::Now()` to compute the value of `deadline` that it passed into us.

This may help avoid the *next* iteration of this loop, but it doesn't really solve the problem of the deadline being computed wrong for the initial iteration.

Reply:
This does solve the issue of getting into an indefinite loop - verified on my repro of the issue. I agree it's weird that the first iteration of the timer can still be computed incorrectly, though, so this also isn't my favorite for that reason.